Library/Homebrew/cask/cask_loader.rb

```diff
@@ -228,18 +228,25 @@ def load(config:)
         json_cask = @from_json || Homebrew::API::Cask.all_casks[token]
         cask_source = JSON.pretty_generate(json_cask)
 
-        json_cask = Homebrew::API.merge_variations(json_cask).deep_symbolize_keys
+        json_cask = Homebrew::API.merge_variations(json_cask).deep_symbolize_keys.freeze
+
+        tap = Tap.fetch(json_cask[:tap]) if json_cask[:tap].to_s.include?("/")
+        tap_git_head = json_cask[:tap_git_head]
 
         # Use the cask-source API if there are any `*flight` blocks or the cask has multiple languages
         if json_cask[:artifacts].any? { |artifact| FLIGHT_STANZAS.include?(artifact.keys.first) } ||
            json_cask[:languages].any?
           cask_source = Homebrew::API::Cask.fetch_source(token, git_head: json_cask[:tap_git_head])
-          return FromContentLoader.new(cask_source).load(config: config)
+          return FromContentLoader.new(cask_source).load(config: config).tap do |cask|
+            # We can't pass these variables normally to `Cask::Cask.new` since we load it from a string.
+            cask.instance_variable_set(:@loaded_from_api, true)
+            cask.instance_variable_set(:@tap, tap)
+            cask.instance_variable_set(:@tap_git_head, tap_git_head)
+          end
```

Comment:
This is kind of a hack but IMO it's better to do it like this than allow access to them by normal means once the cask has been loaded.

Reply:
I feel the formula way to do this is probably better where we don't use the source-downloaded cask at all except at install time (at which point these things shouldn't matter as much). Unless there's cask-specific concerns that prevent that from being possible

It would substantially improve speed of regular cask operations, and allow the actual downloading to only occur when the install downloads are too.

Reply:
Similarly: we should consider using the `url` and `sha256` from the API and ignore the one from the source for improved security.

Reply:
Both good suggestions. I really like the idea of only loading the cask source when necessary which I think would be during installs (we store the cask files after install so we should already have the source necessary for uninstall). I think it might change the artifact output of `Cask::Cask#to_h` though for the `*flight` stanzas we'd have to ignore but that's probably not a big deal since those are just placeholders anyway right now.

I assume this is what you're talking about.

https://github.com/Homebrew/brew/blob/7e48ed50f74e2ef8956ef6ed6078d02c0457c5aa/Library/Homebrew/formula_installer.rb#L1185-L1191

It's honestly probably beyond the scope of this PR so I might just drop this for now and come back to it after figuring that out. I'm not sure if there's one choke point in the cask install/reinstall/upgrade code where we could load the cask source and be fine.

Reply:
Probably just need to put it in `Cask::Installer.initialize`.

Reply:
It overlaps my security PR as well so I might have a look tomorrow if I can.

Reply:
I might see how much we can re-use from the existing code on the formula side too. Could be a hefty refactor if so but it'll be worth it.

Reply:
Feel free to tackle it if you want. Otherwise, I'll get to it when I get to it.

I think really shouldn't be that bad now that I looked at it.

1. Skip loading `flight` and `language` stanzas in the `Cask::CaskLoader::FromAPILoader` (outside of install/uninstall/reinstall/upgrade I can't think of any place other than the info command where these might be missed).
2. Set some sort of flag on the cask like `source_only` when loading it from the API JSON.
3. In `Cask::Installer.initialize` fetch the cask from source using the source API fetcher if it's source only and loaded from the API.

There was also the idea to add a source only flag to the cask JSON so we wouldn't have to do this processing locally after loading from the API.

As an aside the above formula source fetching code should probably eventually end up somewhere in `Homebrew::API`.

Reply:
Agreed with all of this @apainintheneck 👍🏻